Display unique group panel for difficulty grouping

frenzibyte opened the pull request:
![CleanShot 2025-05-07 at 13 15 17](https://github.com/user-attachments/assets/472c3812-1e7c-4bcb-a09b-69453c8eb5fd)

frenzibyte committed:
Update star difficulty panel group with recent design changes

frenzibyte committed:
Display unique group panel for difficulty grouping

Update existing test coverage

frenzibyte commented on osu.Game/Screens/SelectV2/BeatmapCarouselFilterGrouping.cs:
```diff
@@ -139,7 +139,7 @@ void addItem(CarouselItem i)
                     {
                         starGroup = (int)Math.Floor(beatmap.StarRating);
-                        return new GroupDefinition(starGroup + 1, $"{starGroup} - {starGroup + 1} *");
+                        return new GroupDefinition(starGroup + 1, string.Empty);
```
I've removed the title here because it's not used in the star-diff group panel. I can bring it back, update it, and use it as a source for the star-diff group panel if preferred.

peppy replied:
I'd argue it's more correct to leave it for debug/correctness purposes.

frenzibyte commented:
See also [discord](https://discord.com/channels/188630481301012481/1359431017680863302/1369618205940908054).

peppy commented on osu.Game/Screens/SelectV2/BeatmapCarousel.cs:
```diff
@@ -367,9 +367,11 @@ public void Filter(FilterCriteria criteria)
         private readonly DrawablePool<PanelBeatmap> beatmapPanelPool = new DrawablePool<PanelBeatmap>(100);
         private readonly DrawablePool<PanelBeatmapSet> setPanelPool = new DrawablePool<PanelBeatmapSet>(100);
         private readonly DrawablePool<PanelGroup> groupPanelPool = new DrawablePool<PanelGroup>(100);
+        private readonly DrawablePool<PanelGroupStarDifficulty> starsGroupPanelPool = new DrawablePool<PanelGroupStarDifficulty>(100);
```
i'm pretty sure there's a max of 11 panels ever? probably doesn't even need pooling to be honest..

peppy commented on osu.Game/Screens/SelectV2/BeatmapCarousel.cs:
```diff
@@ -397,7 +399,10 @@ protected override Drawable GetDrawableForDisplay(CarouselItem item)
             {
-                case GroupDefinition:
+                case GroupDefinition group:
+                    if (group.Data is int)
```
This is super haphazard. Why not use [`StarDifficulty`](https://github.com/ppy/osu/blob/d37c1bb6d04269f32d9739dcc74fb781bb937358/osu.Game/Beatmaps/StarDifficulty.cs#L9)?

frenzibyte commented:
Addressed reviews above, and also applied a minor change in the stars grouping logic so that it determines the star group after rounding the rating to 2 decimal places. Otherwise beatmaps with 2.00 star rating (rounded from 1.9979797 etc.) could end up in the "1 Stars" group.
